File: questao14/java/Configuracao.java
package questao14.java;

public class Configuracao {
    // Instância única (Singleton)
    private static Configuracao instanciaUnica = null;

    private String host;

    // Construtor privado
    private Configuracao() {
        // Definir valores padrão da configuração
        this.host = "defaultHost";
    }

    // Método público para retornar a única instância
    public static Configuracao getInstancia() {
        if (instanciaUnica == null) {
            instanciaUnica = new Configuracao();
        }
        return instanciaUnica;
    }

    public void setHost(String host) {
        this.host = host;
    }

    public String getHost() {
        return this.host;
    }

    public static void main(String[] args) {
        // Testando o Singleton
        Configuracao config1 = Configuracao.getInstancia();
        Configuracao config2 = Configuracao.getInstancia();

        // Definindo o host na primeira instância
        config1.setHost("localhost");

        // Verificando se a outra instância reflete a alteração
        System.out.println(config2.getHost());  // Deve imprimir 'localhost'

        // Verificando se config1 e config2 são a mesma instância
        System.out.println(config1 == config2);  // Deve imprimir 'true'
    }
}

File: questao14/go/configuracao.go
package main

import (
    "fmt"
    "sync"
)

// Struct Configuracao (Singleton)
type Configuracao struct {
    host string
}

var instanciaUnica *Configuracao
var once sync.Once

// Função para retornar a única instância de Configuracao (Singleton)
func GetInstancia() *Configuracao {
    once.Do(func() {
        instanciaUnica = &Configuracao{host: "defaultHost"}
    })
    return instanciaUnica
}

func main() {
    // Testando o Singleton
    config1 := GetInstancia()
    config2 := GetInstancia()

    // Definindo o host na primeira instância
    config1.host = "localhost"

    // Verificando se a outra instância reflete a alteração
    fmt.Println(config2.host)  // Deve imprimir 'localhost'

    // Verificando se config1 e config2 são a mesma instância
    fmt.Println(config1 == config2)  // Deve imprimir 'true'
}
